### ISSUE TITLE
Diff highlighting broken (looks streaming search related)

### ISSUE THREAD

**Issue by rvantonder** [sourcegraph/sourcegraph#19403]
Looks like a regression with streaming search, because it seems to manifest only with streaming on:

[example](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type:diff+return&patternType=regexp):

<img width="1434" alt="Screen Shot 2021-03-23 at 7 12 22 PM" src="https://user-images.githubusercontent.com/888624/112244035-d9b55280-8c0b-11eb-93e9-56ad4e30f28d.png">

With streaming off:

<img width="1421" alt="Screen Shot 2021-03-23 at 7 11 38 PM" src="https://user-images.githubusercontent.com/888624/112244119-ffdaf280-8c0b-11eb-86a2-b0a6ba741148.png">

Commit message highlighting via `type:commit` appears unaffected. Not sure if backend or frontend so just assigning everyone.

**Comment by keegancsmith** [comment]
Will take a look. I believe this happens without streaming as well sometimes. Need to see if I can find the other issue about this (or maybe I just have it in my org files).

**Comment by keegancsmith** [comment]
This is correctly highlighted in the src-cli + streaming, which makes me think it has something to do with how webapp highlights commits. I think this is an old issue, I do have memory of seeing it before:

![image](https://user-images.githubusercontent.com/187831/112325090-104aa600-8cbc-11eb-9976-5a1b3e8dd7ae.png)

vs

![image](https://user-images.githubusercontent.com/187831/112325150-1ccefe80-8cbc-11eb-8eb3-e0074917ea5e.png)

(match count difference is due to using a lower count: limit when I did the cli query)

**Comment by limitedmage** [comment]
Looking. Also ran into the separate issue: https://sourcegraph.slack.com/archives/C07KZF47K/p1616608611137800

**Comment by limitedmage** [comment]
Root cause: The library we're using to convert Markdown to HTML, Marked, [converts tabs to spaces with **no way to disable this**](https://github.com/markedjs/marked/issues/1559#issuecomment-578509104). So all our highlighting calculations are broken once the tabs have been converted to spaces.

This doesn't happen in non-streaming search because we return HTML of the code instead of Markdown.

**Comment by limitedmage** [comment]
Not sure what the best way to fix this is. @keegancsmith  What is the feasibility of modifying streaming search backend to return HTML like GraphQL search does instead of Markdown?

**Comment by keegancsmith** [comment]
> Not sure what the best way to fix this is. @keegancsmith What is the feasibility of modifying streaming search backend to return HTML like GraphQL search does instead of Markdown?

We can do that. It is a bit strange we have markdown as the response type, really you would just expect a raw byte buffer like we get with line content/etc. So I'd like to better understand how this all works. I would expect we can have a better arch for streaming before we start having 3rd party clients depend on the API. It seems we ask the backend to highlight code blocks, I see requests to graphql asking to highlight a codeblock adn we get back HTML. Is there a better way to structure this overall?

**Comment by stefanhengl** [comment]
@keegancsmith and I discussed this a bit. We think the backend should just send back content as plain text, without the \```COMMIT_MSG or \```diff markers. The frontend wouldn't have to call "Marked" at all. We would still call the backend for highlighting though. 

Only somewhat related, we noticed that sending content back to the backend for highlighting can take a lot of time. Maybe streaming could send down the first N events already highlighted? @limitedmage WDYT?
